src/modules/flow/json/json.c

```diff
@@ -0,0 +1,221 @@
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sol-flow/json.h"
+#include "sol-flow-internal.h"
+
+#include <sol-json.h>
+#include <sol-util.h>
+#include <sol-types.h>
+#include <errno.h>
+
+static const struct sol_blob_type sub_json_blob_type = {
+    .api_version = SOL_BLOB_TYPE_API_VERSION,
+};
+
+struct sol_json_object_key {
+    struct sol_blob *json_object;
+    char *key;
+};
+
+static int
+json_object_get_key_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct sol_json_object_key *mdata = data;
+    const struct sol_flow_node_type_json_object_get_key_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_JSON_OBJECT_GET_KEY_OPTIONS_API_VERSION,
+        -EINVAL);
+    opts = (const struct sol_flow_node_type_json_object_get_key_options *)
+        options;
+
+    mdata->key = strdup(opts->key);
+    SOL_NULL_CHECK(mdata->key, -ENOMEM);
+
+    return 0;
+}
+
+static void
+json_object_get_key_close(struct sol_flow_node *node, void *data)
+{
+    struct sol_json_object_key *mdata = data;
+
+    if (mdata->json_object)
+        sol_blob_unref(mdata->json_object);
+    free(mdata->key);
+}
+
+static struct sol_blob *
+create_sub_json_object(struct sol_blob *parent, struct sol_json_token *token)
+{
+    struct sol_json_scanner scanner;
+    const char *mem;
+
+    mem = token->start;
+    sol_json_scanner_init(&scanner, parent->mem, parent->size);
+    if (!sol_json_scanner_skip_over(&scanner, token))
+        return NULL;
+
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_OBJECT_END) {
```

Comment:
Actually they are slightly different. The only thing in common is the use of  sol_json_scanner_init and sol_json_scanner_skip_over and this is already exported in sol-json.h.
Anyway, I moved the validator function to sol-json.h to be used in converted, because it makes sense to have that helper over there.